### ISSUE TITLE
C api - Program failing during training when compiling with optimizations flag (-O2) or when libtensorflow is built with optimized instructions (AVX SSE), or when keras model is compiled with "sparse_categorical_crossentropy"

### ISSUE THREAD

**Issue by 7PintsOfCherryGarcia** [tensorflow/tensorflow#32288]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):Mixed setup. Keras model is built based on stock example. Training happens in custom C code that adapts an existing TF example
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04, CentOS 7.6 
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Both, binary and source. Keras model is built using python package installed with pip. Training program uses libtensorflow compiled from source
- TensorFlow version (use command below): 2.0
- Python version: 3.6(Ubuntu), 3.7(CentOS)
- Bazel version (if compiling from source): 0.26.1
- GCC/Compiler version (if compiling from source): 7.4(Ubuntu) and 9.1(CentOS)
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
My env:
    [tf_env.txt](https://gist.github.com/7PintsOfCherryGarcia/31247fa9e421bddd158272c01786911e)

**Describe the current behavior**
I am using TF 2.0 [C api](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/c/c_api.h) to recreate the tutorial in [here](https://www.tensorflow.org/beta/tutorials/keras/basic_classification)

Keras model is built using python package as indicated in the tutorial with slight modifications. "categorical_crossentropy" is used as loss function instead of "sparse_categorical_crossentropy" as the latter produces the error referenced in this issue. Model is saved using keras.experimental.export_saved_model.

Model is saved in a folder called "saved_model", **name of this folder** is **hard coded** in C program.

I link against libtensorflow.so.2.0.0 and libtensorflow_framework.so.2.0.0 as compiled with the following command:

`./configure; bazel build -c opt //tensorflow/tools/lib_package:libtensorflow` (no cuda support)

**Describe the expected behavior**
If compiled **without** optimization (-O<num>) program runs normally and model can be trained successfully. As seen can be observed in the output.
If compiled **with** optimization (-O<num>) program fails at training step.
If libtensorflow is **compiled with AVX,SSE** support, program fails at training step.
If keras model is compiled using "sparse_categorical_crossentropy", program fails at training step.
All modes of failure end with the same error (see below)

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
Model is build using 

The steps I follow are:
1) Create keras model and save it to a file.
python [createModel.py](https://gist.github.com/7PintsOfCherryGarcia/094e68585be228e42efc5764a97ab5d7)

2) Download training data and store in two files: data.txt and labels.txt, the names of the input data are **hard coded in C program**. To download data:
python [getData.py](https://gist.github.com/7PintsOfCherryGarcia/455370c9550a6c5c10f9135155a8b282)

3) Run C [program](https://gist.github.com/7PintsOfCherryGarcia/a5935bc0410e64bca2a8f252620eff92), the program will read the saved model, load training data and labels. Make predictions using the untrained model, train the model, run predictions again.
LD_LIBRARY_PATH=/path/to/libtensorflow/lib ./[CFashionMnist](https://gist.github.com/7PintsOfCherryGarcia/a5935bc0410e64bca2a8f252620eff92) The program will print predicted values as well as loss values during training.

Program is compiled with:
For normal execution
  gcc -Wall -I /path/to/libtensorflow/include -L /path/to/libtensorflow/lib -o CFashionMnist CFashionMnist.c -ltensorflow
For failing execution
  gcc -Wall **-O2** -I /path/to/libtensorflow/include -L /path/to/libtensorflow/lib -o CFashionMnist CFashionMnist.c -ltensorflow

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

When compiled with optimizations flags (-O<num>) I get:

2019-09-06 14:16:51.617721: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX
2019-09-06 14:16:51.642262: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 1696290000 Hz
2019-09-06 14:16:51.642842: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x557c6f2b89a0 executing computations on platform Host. Devices:
2019-09-06 14:16:51.642893: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Host, Default Version
2019-09-06 14:16:51.643335: I tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel from: keras_model
2019-09-06 14:16:51.648699: I tensorflow/cc/saved_model/reader.cc:54] Reading meta graph with tags { train }
2019-09-06 14:16:51.662135: I tensorflow/cc/saved_model/loader.cc:202] Restoring SavedModel bundle.
2019-09-06 14:16:51.693512: I tensorflow/cc/saved_model/loader.cc:151] Running initialization op on SavedModel bundle at path: keras_model
2019-09-06 14:16:51.712088: I tensorflow/cc/saved_model/loader.cc:311] SavedModel load for tags { train }; Status: success. Took 68754 microseconds.
**ERROR: Node 'dense_1_target' (type: 'Placeholder', num of outputs: 1) does not have output 2044864112**

If I build libtensorflow with support for optimized instructions I get the same error minus the optimized instructions warning.

**Comment by 7PintsOfCherryGarcia** [comment]
This errors seems to appear randomly. Which makes it very difficult to debug particularly on such an api that is not very informative. 

I am getting the same error whenerver I modify the struct that holds all my graph information

**Comment by 7PintsOfCherryGarcia** [comment]
So I figured it out if anybody is interested. Turns out I'm as dumb as it gets.
When indicating the the index of my "target", I recycled the previous lines of code and just reset the index of "output" to 1 and left "target" index uninitialized:

In Belly_ModelOperations()
  model->output.oper = TF_GraphOperationByName(model->graph, "dense_1/Softmax");
  model->output.index = 0;
  if(model->output.oper == NULL) {
    printf("ERROR output\n");
    return -1;
  }

  model->target.oper = TF_GraphOperationByName(model->graph, "dense_1_target");
  model->output.index = 0; //  <===============BAD
  //Chaneg to:
  model->target.index = 0; //   <==================GOOD
  if(model->target.oper == NULL) {
    printf("ERROR target\n");
    return -1;
  }

I still have no clue why not initializing the target index ended up in error on certain occasions
